### ISSUE TITLE
SEGV - READ memory access

### ISSUE THREAD

**Issue by RootUp** [vim/vim#5511]
**VIM Version:**
```
$ ./vim --version
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jan 19 2020 23:05:46)
Included patches: 1-131
Compiled by input0@zero
Huge version without GUI.  Features included (+) or not (-):
+acl               -farsi             -mouse_sysmouse    -tag_old_static
+arabic            +file_in_path      +mouse_urxvt       -tag_any_white
+autocmd           +find_in_path      +mouse_xterm       -tcl
+autochdir         +float             +multi_byte        +termguicolors
-autoservername    +folding           +multi_lang        +terminal
-balloon_eval      -footer            -mzscheme          +terminfo
+balloon_eval_term +fork()            +netbeans_intg     +termresponse
-browse            +gettext           +num64             +textobjects
++builtin_terms    -hangul_input      +packages          +textprop
+byte_offset       +iconv             +path_extra        +timers
+channel           +insert_expand     -perl              +title
+cindent           +job               +persistent_undo   -toolbar
+clientserver      +jumplist          +popupwin          +user_commands
+clipboard         +keymap            +postscript        +vartabs
+cmdline_compl     +lambda            +printer           +vertsplit
+cmdline_hist      +langmap           +profile           +virtualedit
+cmdline_info      +libcall           -python            +visual
+comments          +linebreak         -python3           +visualextra
+conceal           +lispindent        +quickfix          +viminfo
+cryptv            +listcmds          +reltime           +vreplace
+cscope            +localmap          +rightleft         +wildignore
+cursorbind        -lua               -ruby              +wildmenu
+cursorshape       +menu              +scrollbind        +windows
+dialog_con        +mksession         +signs             +writebackup
+diff              +modify_fname      +smartindent       +X11
+digraphs          +mouse             -sound             +xfontset
-dnd               -mouseshape        +spell             -xim
-ebcdic            +mouse_dec         +startuptime       -xpm
+emacs_tags        -mouse_gpm         +statusline        +xsmp_interact
+eval              -mouse_jsbterm     -sun_workshop      +xterm_clipboard
+ex_extra          +mouse_netterm     +syntax            -xterm_save
+extra_search      +mouse_sgr         +tag_binary        
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/usr/local/share/vim"
Compilation: afl-clang-fast -c -I. -Iproto -DHAVE_CONFIG_H     -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: afl-clang-fast   -L/usr/local/lib -Wl,--as-needed -o vim    -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinfo -lelf -lnsl  -ldl
```

**MData:**
```
While fuzzing VIM with enabling different mode a read memory access is observed (SEGV).
Address point to zero page, however I was not able to control the register for further exploitation.
```

**GBD BT:**
```
$ gdb -q ./vim

warning: ~/peda/peda.py: No such file or directory
Reading symbols from ./vim...done.
(gdb) r -u NONE -X -Z -e -s -S poc1 -c ':qa!'
Starting program: /home/input0/vim/src/vim -u NONE -X -Z -e -s -S poc1 -c ':qa!'
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
__strlen_avx2 () at ../sysdeps/x86_64/multiarch/strlen-avx2.S:62
62	../sysdeps/x86_64/multiarch/strlen-avx2.S: No such file or directory.
(gdb) bt 
#0  __strlen_avx2 () at ../sysdeps/x86_64/multiarch/strlen-avx2.S:62
#1  0x00000000004b1f72 in bracketed_paste (mode=<optimized out>, drop=<optimized out>, gap=<optimized out>) at edit.c:4945
#2  0x0000000000548392 in getcmdline_int (firstc=<optimized out>, count=<optimized out>, indent=<optimized out>, init_ccline=<optimized out>)
    at ex_getln.c:2266
#3  0x000000000067b795 in nv_search (cap=0x7fffffffa628) at normal.c:4250
#4  0x000000000066f356 in normal_cmd (oap=<optimized out>, toplevel=<optimized out>) at normal.c:1071
#5  0x00000000005374c2 in exec_normal (was_typed=0, use_vpeekc=0, may_use_terminal_loop=0) at ex_docmd.c:7675
#6  0x0000000000536e55 in exec_normal_cmd (cmd=<optimized out>, remap=<optimized out>, silent=0) at ex_docmd.c:7665
#7  ex_normal (eap=0x7fffffffa8a0) at ex_docmd.c:7591
#8  0x00000000005249f0 in do_one_cmd (sourcing=<optimized out>, cstack=<optimized out>, cmdlinep=<optimized out>, fgetline=<optimized out>, 
    cookie=<optimized out>) at ex_docmd.c:2482
#9  do_cmdline (cmdline=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>, flags=<optimized out>) at ex_docmd.c:975
#10 0x00000000007a68f9 in do_source (fname=<optimized out>, check_other=<optimized out>, is_vimrc=<optimized out>) at scriptfile.c:1339
#11 0x00000000007a54d9 in cmd_source (fname=0xc54ec3 "poc1", eap=<optimized out>) at scriptfile.c:933
#12 0x00000000005249f0 in do_one_cmd (sourcing=<optimized out>, cstack=<optimized out>, cmdlinep=<optimized out>, fgetline=<optimized out>, 
    cookie=<optimized out>) at ex_docmd.c:2482
#13 do_cmdline (cmdline=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>, flags=<optimized out>) at ex_docmd.c:975
#14 0x00000000009794a9 in exe_commands (parmp=<optimized out>) at main.c:3135
#15 vim_main2 () at main.c:795
#16 0x00000000009763d5 in main (argc=<optimized out>, argv=<optimized out>) at main.c:444
(gdb) i r
rax            0x0	0
rbx            0x0	0
rcx            0x0	0
rdx            0x0	0
rsi            0x0	0
rdi            0x0	0
rbp            0x0	0x0
rsp            0x7fffffffa288	0x7fffffffa288
r8             0x0	0
r9             0x0	0
r10            0xfffffffffffffffc	-4
r11            0x0	0
r12            0xfffffffffffffffc	-4
r13            0x0	0
r14            0xffffffff	4294967295
r15            0xfffffffffffffffc	-4
rip            0x7ffff69e25a1	0x7ffff69e25a1 <__strlen_avx2+17>
eflags         0x10283	[ CF SF IF RF ]
cs             0x33	51
ss             0x2b	43
ds             0x0	0
es             0x0	0
fs             0x0	0
gs             0x0	0
(gdb) 
```

**ASAN BT:**
```
ASAN:DEADLYSIGNAL
=================================================================
==28777==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x7fa9543485a1 bp 0x7ffeb8568d90 sp 0x7ffeb8568518 T0)
==28777==The signal is caused by a READ memory access.
==28777==Hint: address points to the zero page.
ASAN:DEADLYSIGNAL
AddressSanitizer: nested bug in the same thread, aborting.
```

**To Reproduce:** vim -u NONE -X -Z -e -s -S [poc](https://github.com/vim/vim/files/4087239/poc.zip) -c ':qa!'

**Comment by dpelle** [comment]
I can reproduce it with:
```
$ vim --clean -e -s -c 'exe "norm /\x80PS"'
```

**Comment by tcullum-rh** [comment]
@janlazo I get this on `if (STRLEN(end) >= NUMBUFLEN)` because `end` is `NULL` here and STRLEN is called on it. This seems to occur before the patched area. See screenshot below:

![vim_strlen](https://user-images.githubusercontent.com/64049057/79928345-8e8d3100-83f7-11ea-9302-aaccfc046063.png)

Used same payload as @dpelle .

**Comment by chrisbra** [comment]
@tcullum-rh you get a crash? can you please post a gdb backtrace using the `bt`  command? just copy paste the output, no screenshots necessary. How did you reproduce this?

**Comment by dpelle** [comment]
To @tcullum-rh 

In edit.c of vim 8.2.0616:, I see:
```
4944     // If the end code is too long we can't detect it, read everything.
4945     if (end != NULL && STRLEN(end) >= NUMBUFLEN)
4946         end = NULL;
```
But on your screenshot I see:
```
 if (STRLEN(end) >= NUMBUFLEN)
```
So it looks like you're not using the latest vim with the fix.
The bug was fixed in vim-8.2.0135.
Which version of vim are you using?  You can use the `:version` command.

**Comment by tcullum-rh** [comment]
@chrisbra and @dpelle no action is needed here. I was confused because I hadn't seen this patch: https://github.com/vim/vim/commit/fe4bbac1166f2e4e3fa18cb966ec7305198c8176

The confusion was because I was told (elsewhere) that https://github.com/vim/vim/commit/98a336dd497d3422e7efeef9f24cc9e25aeb8a49 was the commit which solved the above bug and I didn't understand how this was the case because the debugger was showing that the issue was resulting from the one patched in https://github.com/vim/vim/commit/fe4bbac1166f2e4e3fa18cb966ec7305198c8176 . When looking at it, it made no sense to me how that first commit could have solved the NULL ptr issue that I was seeing in the debugger.

Yes, I'm working with a previous Vim version and I just tested the current Vim, and the issue doesn't happen there. Thank you and I apologize for any confusion.
